src/cobra/core/reaction.py

```diff
@@ -224,65 +329,117 @@ def update_variable_bounds(self):
+
+        When using a `HistoryManager` context, this attribute can be set
+        temporarily, reversed when the exiting the context.
+
+        Raises
+        ------
+        ValueError
+            If lower bound higher than the current upper bound. via _check_bounds.
+
+        See Also
+        --------
+        _check_bounds
+        """
         # Validate bounds before setting them.
         self._check_bounds(value, self._upper_bound)
         self._lower_bound = value
         self.update_variable_bounds()
 
     @property
-    def upper_bound(self):
-        """Get or set the upper bound
-
-        Setting the upper bound (float) will also adjust the associated optlang
-        variables associated with the reaction. Infeasible combinations,
-        such as a upper bound lower than the current lower bound will
-        update the other bound.
+    def upper_bound(self) -> float:
+        """Get the upper bound.
 
-        When using a `HistoryManager` context, this attribute can be set
-        temporarily, reversed when the exiting the context.
+        Returns
+        -------
+        float
+            The upper bound of the reaction.
         """
         return self._upper_bound
 
     @upper_bound.setter
     @resettable
-    def upper_bound(self, value):
+    def upper_bound(self, value: float) -> None:
+        """Set the upper bound.
+
+        Parameters
+        ----------
+        value: float
+            The value to set the upper bound.
+
+        Setting the upper bound (float) will also adjust the associated optlang
+        variables associated with the reaction.
+
+        When using a `HistoryManager` context, this attribute can be set
+        temporarily, reversed when the exiting the context.
+
+        Raises
+        ------
+        ValueError
+            If upper bound lower than the current upper bound. via _check_bounds.
+
+        See Also
+        --------
+        _check_bounds
+        """
         # Validate bounds before setting them.
         self._check_bounds(self._lower_bound, value)
         self._upper_bound = value
         self.update_variable_bounds()
 
     @property
-    def bounds(self):
-        """Get or set the bounds directly from a tuple
+    def bounds(self) -> Tuple[float, float]:
+        """Get or the bounds.
 
-        Convenience method for setting upper and lower bounds in one line
-        using a tuple of lower and upper bound. Invalid bounds will raise an
-        AssertionError.
-
-        When using a `HistoryManager` context, this attribute can be set
-        temporarily, reversed when the exiting the context.
+        Returns
+        -------
+        tuple: lower_bound, upper_bound
+            A tuple of floats, representing the lower and upper bound.
         """
         return self.lower_bound, self.upper_bound
 
     @bounds.setter
     @resettable
-    def bounds(self, value):
+    def bounds(self, value: Union[Tuple[float, float], List[float]]) -> None:
```

Comment:
It would be more concise to say that a [Sequence](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) interface needs to be supported.

```suggestion
    def bounds(self, value: Sequence[float]) -> None:
```

but if you want to highlight that the usual case is a pair that's fine, too:

```suggestion
    def bounds(self, value: Union[Tuple[float, float], Sequence[float]]) -> None:
```